data-contracts tutorial

sandonjacobs opened the pull request:
Tutorial explaining how to manage schemas with Confluent Schema registry, using the Confluent Terraform Provider and Gradle. Code examples show how to configure producers and consumers to use specific versions of the schema. The build also includes data migration rules to illustrate our support for o"breaking" schema evolution changes.

Also included is an update of the common module to use Java 17.

sandonjacobs committed:
Initial setup for data contracts with terraform module.

sandonjacobs committed:
terraform setup with v1 of schema.

This includes a major_version compatibility group.

sandonjacobs committed:
Adding v2 of membership schema.

sandonjacobs committed:
Managing schema registration with gradle.

sandonjacobs committed:
Updates to documentation.

sandonjacobs committed:
documenation, module rename, all gradle for schema registration and download.

sandonjacobs committed:
typos and additional instructions.

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+
+</details>
+
+Running this application will print the events being consumed from Kafka:
```
Is there a gradle command to run it? or instruction to run from IntelliJ? I tried opening the `data-contracts` directory in intellij but it's not finding the `shared` module:
![image](https://github.com/user-attachments/assets/556a9d0f-ae8a-487f-9f3d-1ed102806750)
I think that `settings.gradle.kts` need some `include` directives to avoid these errors, but wondering if there is something different about your IntelliJ setup that avoids them

davetroiano replied:
i would remove the `details` section and just add one sample:

Running this application will log the events being consumed from Kafka, e.g.:
```plaintext
Received Membership 8e65d8e2-4ad8-475f-8f74-d724865ddbd8, {"user_id": "8e65d8e2-4ad8-475f-8f74-d724865ddbd8", "start_date": "2022-06-13", "end_date": "2025-03-25"}
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+  done > ../shared/src/main/resources/confluent.properties
+```
+
+All Kafka Client code in this project loads connection properties form `shared/src/main/resources/confluent.properties`. For an example of this 
````
```suggestion
All Kafka Client code in this project loads connection properties from `shared/src/main/resources/confluent.properties`. For an example of this 
```

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+# Managing Data Contracts
+
+Data contracts consist not only of the schemas to define the structure of events, but also rulesets allowing for more fine-grained validations,
+controls, and discovery. In this tutorial, we'll evolve a schemas and add data quality and migration rules.
```
```suggestion
controls, and discovery. In this tutorial, we'll evolve a schema and add data quality and migration rules.
```

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+
+In the workflow above, we see these tools in action:
+* The [Confluent Terraform Provider](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs) is used to define Confluent Cloud assets (Kafka cluster(s), Data Governance, Kafka Topics, and Schema Configurations).
+* Using the newly created Schema Registry, data engineers and architects define the schema of the events that comprise the organization's canonical data model - i.e. entities, events, and commands that are shared across applications. - along with other parts of the data contract. This includes data quality rules, metadata, and migration rules. A gradle plugin is utilized to register the schemas a ...
```
```suggestion
* Using the newly created Schema Registry, data engineers and architects define the schema of the events that comprise the organization's canonical data model (i.e., entities, events, and commands that are shared across applications) along with other parts of the data contract. This includes data quality rules, metadata, and migration rules. A Gradle plugin is utilized to register the schemas and related elements of the data contract with Schema Registry.
```

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+along with metadata and migration rules. 
+
+We will evolve the `membership` schema, refactoring the events to encapsulate the date-related fields of version 1 into its
+own `record` type in version 2. Typically this would be a breaking change. However, [data migration rules](https://docs.confluent.io/cloud/current/sr/fundamentals/data-contracts.html#migration-rules) in the schema registry
```
```suggestion
own `record` type in version 2. Typically this would be a breaking change. However, [data migration rules](https://docs.confluent.io/cloud/current/sr/fundamentals/data-contracts.html#migration-rules) in Schema Registry
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+}
+```
+
+Then we configure the subject containing the value of events for this topic using `confluent_subject_config`. This tells the Schema Registry that `membership-avro-value` schemas will be `BACKWARD` compatible and defines an attribute called `major_version` to be used as the `compatibilityGroup`.
````
```suggestion
Then we configure the subject containing the value of events for this topic using `confluent_subject_config`. This tells Schema Registry that `membership-avro-value` schemas will be `BACKWARD` compatible and defines an attribute called `major_version` to be used as the `compatibilityGroup`.
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+}
+```
+
+Using the [schema-registry-plugin](https://github.com/ImFlog/schema-registry-plugin) for gradle, we can register this schema with the Confluent Schema Registry.
````
```suggestion
Using the [schema-registry-plugin](https://github.com/ImFlog/schema-registry-plugin) for Gradle, we can register this schema with Schema Registry.
```

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+In the workflow above, we see these tools in action:
+* The [Confluent Terraform Provider](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs) is used to define Confluent Cloud assets (Kafka cluster(s), Data Governance, Kafka Topics, and Schema Configurations).
+* Using the newly created Schema Registry, data engineers and architects define the schema of the events that comprise the organization's canonical data model - i.e. entities, events, and commands that are shared across applications. - along with other parts of the data contract. This includes data quality rules, metadata, and migration rules. A gradle plugin is utilized to register the schemas a ...
+* Applications which producer and/or consume these event types can download the schemas from the Schema Registry. In our example, this is a JVM application built using Gradle. A gradle plugin is used to download the schemas, after which another gradle plugin is used to generate Java classes from those schemas - thus providing the application with compile-time type safety.
```
```suggestion
* Applications which produce and/or consume these event types can download the schemas from Schema Registry. In our example, this is a JVM application built using Gradle. A Gradle plugin is used to download the schemas, after which another Gradle plugin is used to generate Java classes from those schemas - thus providing the application with compile-time type safety.
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+./gradlew :data-contracts:schemas:registerSchemasTask
+```
+
+To use this schema in an application - `app-schema-v1` - we utilize the same gradle plugin's `downloadSchemasTask`. First we define the subjects we want to download. In this example, we copy the schema to directory `src/main/avro`:
````
```suggestion
To use this schema in an application - `app-schema-v1` - we utilize the same Gradle plugin's `downloadSchemasTask`. First we define the subjects we want to download. In this example, we copy the schema to directory `src/main/avro`:
```

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+
+With the schema(s) downloaded, we next want to generate Java code to provide our application an SDK with compile-time bindings to serialize and deserialize events to and from the `membership-avro` topic. Let's use the [gradle-avro-dependency-plugin](https://github.com/bakdata/gradle-avro-dependency-plugin) to generate code, using the `generateAvroJava` task.
+
+For simplicity, I have encapsulated `downloadSchemasTask` with the `generateAvroJava` task into a custom gradle task named `generateCode`, executed as follows:
```
imo first person singular looks off in a CD tutorial. avoid or use first plural
```suggestion
For simplicity, the `downloadSchemasTask` and `generateAvroJava` tasks are encapsulated into a custom Gradle task named `generateCode`, executed as follows:
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+}
+```
+
+Register the new schema using our gradle plugin:
````
```suggestion
Register the new schema using our Gradle plugin:
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+}
+```
+
+For more on these data contract configuration enhancements, refer to the CC docs on [data contracts](https://docs.confluent.io/cloud/current/sr/fundamentals/data-contracts.html#configuration-enhancements). 
````
```suggestion
For more on these data contract configuration enhancements, refer to the Confluent Cloud documentation on [Data Contracts](https://docs.confluent.io/cloud/current/sr/fundamentals/data-contracts.html#configuration-enhancements). 
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+```
+
+All Kafka Client code in this project loads connection properties form `shared/src/main/resources/confluent.properties`. For an example of this 
+properties file, see [confluent.properties.orig](shared/src/main/resources/confluent.properties.orig).
````
full URL so that it works on CD
```suggestion
properties file, see [confluent.properties.orig](https://raw.githubusercontent.com/confluentinc/tutorials/refs/heads/master/data-contracts/shared/src/main/resources/confluent.properties.orig).
```

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+))
+```
+
+This implementation must pass the configuration parameters required to serialize the key and value of events to the `membership-avro` topic. In this use case, the key is a `String` and the value is of type `Membership` - generated from the schema. The value uses the schema-registry-aware `Serializer` implementation `KafkaAvroSerializer`. The underlying serializer is also configured to NOT use the ...
````
```suggestion
This implementation must pass the configuration parameters required to serialize the key and value of events to the `membership-avro` topic. In this use case, the key is a `String` and the value is of type `Membership` - generated from the schema. The value uses the schema-registry-aware `Serializer` implementation `KafkaAvroSerializer`. The underlying serializer is also configured to NOT use the latest version of the schema (`use.latest.version`) and to relax the strict compatibility checks (`latest.compatibility.strict`). For more on these settings, see the Confluent Cloud documentation's explanation of [the differences between preregistered and client-derived schemas](https://docs.confluent.io/platform/current/schema-registry/fundamentals/serdes-develop/index.html#handling-differences-between-preregistered-and-client-derived-schemas).
```

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+A closer look at the configuration of this consumer shows we are now using `latest.compatibility.strict` and specifying `major_version=2`
+in the `use.latest.with.metadata` configuration for the deserializer.
+
+Running the `main` function in `ApplicationV2Main` will consume the events from the `membership-avro` topic, but with a noticeable difference from the `app-schama-v1` application:
```
similar to above - I wasn't actually able to run this due to the error I hit in IntelliJ

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+
+We'll create producer and consumer classes to configure and provide Kafka clients for our application(s). In the `shared` module there are implementations to encapsulate this behavior:
+
+```kotlin
````
thinking of ways to trim down the README and think we don't need the code snippet. say that the clients are dummy implementations, e.g., the client simply logs events, but dont need to include the code in a README

sandonjacobs replied:
Would making these "collapsible" work a lil better?

davetroiano replied:
my pref would be a link (non-relative for out friend CD) to the source for this particular one since the producer / consumer are kind of background noise far from the point of this tutorial. my high level 2c on snippets by example - definitely do show a snippet and explain the client configs that are relevant to the tutorial topic (`LATEST_COMPATIBILITY_STRICT`, `USE_LATEST_VERION`, `USE_LATEST_WITH_METADATA`), but inlining code snippets of the class hierarchy and showing a consumer that logs is too much detail and not relevant enough to make the README cut IMO... the snippets kinda detract from the thing to highlight which is those three configs

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+
+Now implement classes to produce and consume events with this schema. First the producer class, extending `BaseProducer`:
+
+```kotlin
+class MembershipProducer: BaseProducer<String, Membership>(mapOf(
+    ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG to "org.apache.kafka.common.serialization.StringSerializer",
+    ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG to "io.confluent.kafka.serializers.KafkaAvroSerializer",
+    ProducerConfig.CLIENT_ID_CONFIG to "membership-producer-app-v1",
+    AbstractKafkaSchemaSerDeConfig.LATEST_COMPATIBILITY_STRICT to true,
+    AbstractKafkaSchemaSerDeConfig.USE_LATEST_VERSION to false,
+    AbstractKafkaSchemaSerDeConfig.USE_LATEST_WITH_METADATA to "major_version=1"
+))
````
i think the code snippets can go away, or be trimmed down to the most important part - last 3 lines

sandonjacobs replied:
@davetroiano Do you think I should remove the references in the paragraph below to the Serializers?

davetroiano replied:
I could go either way on that... it doesn't seem necessary to cover the tutorial topic but it's close enough to the topic that I think it's alright to leave in there

davetroiano commented on data-contracts/README.md:
```diff
@@ -0,0 +1,439 @@
+
+To exercise these producer and consumer implementations, we created a `main` function in the `ApplicationMain` class to start a consumer instance and a producer instance periodically send random events to the `membership-avro` topic.
+
+<details>
```
i'd remove this. just reference the code with a link and say what it does in English, but no need to put code inline.

davetroiano commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+
+In the `app-schema-v2` module, we'll find a new implementation of a `MembershipConsumer`:
+
+```kotlin
````
similar feedback to above - suggest removing or trimming the code snippets and console output down. it's a little overwhelming and I think distracts from the flow of running through it.

sandonjacobs committed:
mkdir for schema-registry-plugin output.

Addressing PR comment: https://github.com/confluentinc/tutorials/pull/80#discussion_r1919165653

sandonjacobs committed:
Adding instructions to clone repo.

Addresses PR comment: https://github.com/confluentinc/tutorials/pull/80#discussion_r1919165653

sandonjacobs committed:
Correction for terraform output command.

Addressing https://github.com/confluentinc/tutorials/pull/80#discussion_r1919169371

sandonjacobs commented on data-contracts/README.md:
````diff
@@ -0,0 +1,439 @@
+}
+```
+
+Register the new schema using our gradle plugin:
+
+```shell
+./gradlew :data-contracts:schemas:registerSchemasTask
+```
````
Good point, lemme remove this second registration.

sandonjacobs committed:
Removed redundant registerSchemasTask command.

sandonjacobs committed:
Shortening README by removing some code samples.

sandonjacobs committed:
title change.

davetroiano commented:
closing as discussed with @sandonjacobs
